cpp/src/utilities/error_utils.hpp

```diff
@@ -0,0 +1,65 @@
+#ifndef ERRORUTILS_HPP
+#define ERRORUTILS_HPP
+
+#include <cuda.h>
+#include <cuda_runtime_api.h>
+#include <iostream>
+
+#define CUDA_TRY(call)                                                    \
+  {                                                                       \
+    cudaError_t cudaStatus = call;                                        \
+    if (cudaSuccess != cudaStatus) {                                      \
+      std::cerr << "ERROR: CUDA Runtime call " << #call << " in line "    \
+                << __LINE__ << " of file " << __FILE__ << " failed with " \
+                << cudaGetErrorString(cudaStatus) << " (" << cudaStatus   \
+                << ").\n";                                                \
+      return GDF_CUDA_ERROR;                                              \
+    }                                                                     \
+  }
+
+#define RMM_TRY(x) \
+  if ((x) != RMM_SUCCESS) return GDF_MEMORYMANAGER_ERROR;
+
+#define RMM_TRY_CUDAERROR(x) \
+  if ((x) != RMM_SUCCESS) return cudaPeekAtLastError();
+
+#define CUDA_CHECK_LAST() CUDA_TRY(cudaPeekAtLastError())
+
+#define GDF_REQUIRE(F, S) \
+  if (!(F)) return (S);
+
+namespace cudf {
+/**---------------------------------------------------------------------------*
+ * @brief Exception thrown when logical precondition is violated.
+ *
+ * This exception should not be thrown directly and is instead thrown by the
+ * CUDF_EXPECTS macro.
+ *
+ *---------------------------------------------------------------------------**/
+struct logic_error : public std::logic_error {
+  logic_error(char const* const message) : std::logic_error(message) {}
+
+  // TODO Add an error code member? This would be useful for translating an
+  // exception to an error code in a pure-C API
+};
+}  // namespace cudf
+
+#define STRINGIFY_DETAIL(x) #x
+#define CUDF_STRINGIFY(x) STRINGIFY_DETAIL(x)
+
+/**---------------------------------------------------------------------------*
+ * @brief Error checking macro that throws an exception when a condition is
+ * violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw cudf::logic_error if the condition evaluates to false.
+ *---------------------------------------------------------------------------**/
+#define CUDF_EXPECTS(cond, reason)              \
+  (!!(cond)) ? static_cast<void>(0)             \
+             : throw cudf::logic_error(         \
+                   "cuDF failure at: " __FILE__ \
+                   ":" CUDF_STRINGIFY(__LINE__) ". Reason: " reason)
```

Comment:
Is it necessary to have "Reason: " in there? What does GSL do?

Reply:
It's to provide additional context as to why it's expected that the precondition should be true, e.g.,

`CUDF_EXPECTS( lhs->dtype == rhs->dtype, "Column types should match.");`

My thought is that it would provide additional detail, especially when it gets propagated up to the Python level. Otherwise they will just see:

```
cuDF failure at: ../tests/error/error_handling_test.cu:44.
```

Reply:
`gsl` does not provide the equivalent of the `Reason`.

Reply:
:) That's not exactly what I was asking. I was asking if we can reduce output verbosity by eliminating "Reason: " (the actual text) from the output. If users of your macro are writing good, clear code, they should provide reason strings that are clearly reasons for failure. So saying "Reason: " is redundant.

e.g. `cudf failure at: <FILE:LINE>: Column type mismatch`  

Note this is a better reason string than "Column types should match." because it says clearly what went wrong, not what "should" be true.

Reply:
Understood. And done.